java/ant.hints/src/org/netbeans/modules/ant/hints/errors/EnablePreviewAntProj.java

```diff
@@ -18,94 +18,66 @@
-import org.netbeans.api.java.source.CompilationInfo;
 import org.netbeans.api.project.FileOwnerQuery;
-import org.netbeans.modules.java.hints.spi.ErrorRule;
-import org.netbeans.spi.editor.hints.ChangeInfo;
-import org.netbeans.spi.editor.hints.Fix;
 import org.openide.filesystems.FileObject;
-import org.openide.util.NbBundle;
 import org.openide.util.Parameters;
 import org.netbeans.api.project.Project;
 import org.netbeans.api.project.ProjectManager;
+import org.netbeans.modules.java.hints.spi.preview.PreviewEnabler;
 import org.netbeans.spi.project.support.ant.AntProjectHelper;
 import org.openide.util.EditableProperties;
 import org.openide.util.Mutex;
 import org.openide.util.MutexException;
+import org.openide.util.lookup.ServiceProvider;
 
 /**
  * Handle error rule "compiler.err.preview.feature.disabled.plural" and provide
  * the fix for Ant type project.
  *
  * @author arusinha
  */
-public class EnablePreviewAntProj implements ErrorRule<Void> {
+public class EnablePreviewAntProj implements PreviewEnabler {
 
-    private static final Set<String> ERROR_CODES = new HashSet<String>(Arrays.asList(
-            "compiler.err.preview.feature.disabled",          // NOI18N
-            "compiler.err.preview.feature.disabled.plural")); // NOI18N
     private static final String ENABLE_PREVIEW_FLAG = "--enable-preview";   // NOI18N
+    private static final String JAVAC_SOURCE = "javac.source"; // NOI18N
+    private static final String JAVAC_TARGET = "javac.target"; // NOI18N
     private static final String JAVAC_COMPILER_ARGS = "javac.compilerargs"; // NOI18N
     private static final String RUN_JVMARGS = "run.jvmargs"; // NOI18N
 
-    @Override
-    public Set<String> getCodes() {
-        return Collections.unmodifiableSet(ERROR_CODES);
+    private final Project prj;
+
+    public EnablePreviewAntProj(Project prj) {
+        this.prj = prj;
     }
 
     @Override
-    @NonNull
-    public List<Fix> run(CompilationInfo compilationInfo, String diagnosticKey, int offset, TreePath treePath, Data<Void> data) {
+    public void enablePreview(String newSourceLevel) throws Exception {
+        new ResolveAntFix(prj).enablePreview(newSourceLevel);
+    }
 
-        if (SourceVersion.latest() != compilationInfo.getSourceVersion()) {
-            return Collections.<Fix>emptyList();
-        }
+    @ServiceProvider(service=Factory.class, position=1000)
+    public static final class FactoryImpl implements Factory {
 
-        final FileObject file = compilationInfo.getFileObject();
-        Fix fix = null;
-        if (file != null) {
+        @Override
+        public PreviewEnabler enablerFor(FileObject file) {
             final Project prj = FileOwnerQuery.getOwner(file);
 
             if (isAntProject(prj)) {
-                fix = new EnablePreviewAntProj.ResolveAntFix(prj);
+                return new EnablePreviewAntProj(prj);
             } else {
-                fix = null;
+                return null;
```

Comment:
does it make sense to return some NullObject here ?

Reply:
Well, there must a value saying "this provider is not interested". I could invent a special hardcoded value for that, but isn't there already a value that can carry the meaning? I.e. `null`?

Reply:
ok